File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DesignSystem/StatesLiveInStateFolder.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class StatesLiveInStateFolder: QuickSpec {
    override class func spec() {
        Given("A State type in an atomic component") {
            let enums = WoofHarmonize.atomicDesignPackage.enums().withSuffix("State").withoutFolder("_Tokens")
            let structs = WoofHarmonize.atomicDesignPackage.structs().withSuffix("State").withoutFolder("_Tokens")
                .filter { !$0.inheritanceTypesNames.contains("View") }

            Then("It lives in a State folder next to its component") {
                enums.assertTrue(rule: rule) { $0.filePathString.contains("/State/") }
                structs.assertTrue(rule: rule) { $0.filePathString.contains("/State/") }
            }
        }
    }

    private static let rule = Rule(
        description: "Component states live in a State folder next to their component.",
        rationale: "A state is part of a component's API (ButtonState, OverflowMenuState); keeping it next to the component makes the API discoverable.",
        fixHint: "Move the state into <Component>/State/.",
        badExample: "// Molecules/Button/ButtonState.swift\nenum ButtonState { ... }",
        goodExample: "// Molecules/Button/State/ButtonState.swift\nenum ButtonState { ... }"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/designsystem/StatesLiveInStateFolder.kt
package com.perrystreet.woof.konsist.designsystem

import com.lemonappdev.konsist.api.ext.list.withNameEndingWith
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class StatesLiveInStateFolder : BehaviorSpec() {
    init {
        Given("A State type in an atomic component") {
            val states = KonsistUtils.atomicDesignModule.classes()
                .withNameEndingWith("State")
                .filter { it.isTopLevel && !it.path.contains("/_tokens/") }

            Then("It lives in a state package next to its component") {
                states.assertTrue(message = Message) { it.resideInPackage("..state") }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Component states live in a state package next to their component.",
            why = "A state is part of a component's API (ButtonState, OverflowMenuState); keeping it next to the component makes the API discoverable.",
            howToFix = "Move the state into <component>/state/.",
            badExample = "// molecules/button/ButtonState.kt\nenum class ButtonState { ... }",
            goodExample = "// molecules/button/state/ButtonState.kt\nenum class ButtonState { ... }",
        )
    }
}
